cloudflare/resource_cloudflare_custom_ssl.go

```diff
@@ -0,0 +1,284 @@
+package cloudflare
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/cloudflare/cloudflare-go"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/pkg/errors"
+)
+
+func resourceCloudflareCustomSsl() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceCloudflareCustomSslCreate,
+		Read:   resourceCloudflareCustomSslRead,
+		Update: resourceCloudflareCustomSslUpdate,
+		Delete: resourceCloudflareCustomSslDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceCloudflareCustomSslImport,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"zone_id": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"custom_ssl_priority": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"priority": {
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
+					},
+				},
+			},
+			"custom_ssl_options": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"certificate": {
+							Type:     schema.TypeString,
+							Optional: false,
+						},
+						"private_key": {
+							Type:     schema.TypeString,
+							Optional: false,
+						},
+						"bundle_method": {
+							Type:     schema.TypeString,
+							Optional: true,
```

Comment:
As the API has restrictions on what values can be used here, it's helpful to provide validation before the HTTP interactions are made.

```suggestion
							Optional: true,
				                        ValidateFunc: validation.StringInSlice([]string{"ubiquitous", "optimal", "force"}, false),
```

Reply:
I was planning on adding validations as a separate ticket. This one establishes minimal functionality for creating, updating, deleting `custom_ssl` resource. I can add validation for requisite parameters on `POST` , `PUT` encapsulating functions though no problem.

Reply:
Understandable. If we can, I prefer to have as much validation up front as (even with MVP style PRs) to speed up the feedback loop that users get. For example, our Terraform configuration is across ~40 domains and if we apply changes we can wait up to a couple of minutes for the payload to hit the Cloudflare API only to inform us that we've incorrectly formatted a field or gave it an unexpected value. Compared to schema based validation, we get that almost instantly. It does make the user UX quite a bit nicer and doesn't waste as much time on larger installs 🙂

Reply:
Forgot to update this after I added Validation for post fields that take on a set of string values.